grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipEdge.java

```diff
@@ -62,8 +63,15 @@
     private final Cache<Role> valueRole = new Cache<>(Cacheable.concept(), () -> edge().tx().getSchemaConcept(LabelId.of(
             edge().property(Schema.EdgeProperty.RELATIONSHIP_ROLE_VALUE_LABEL_ID))));
 
-    private final Cache<Thing> owner = new Cache<>(Cacheable.concept(), () -> edge().tx().factory().buildConcept(edge().source()));
-    private final Cache<Thing> value = new Cache<>(Cacheable.concept(), () -> edge().tx().factory().buildConcept(edge().target()));
+    private final Cache<Thing> owner = new Cache<>(Cacheable.concept(), () -> {
+        VertexElement vertexElement = edge().source().orElseThrow(() -> GraknTxOperationException.missingOwner(getId()));
+        return edge().tx().factory().<Thing>buildConcept(vertexElement).orElseThrow(() -> GraknTxOperationException.missingOwner(getId()));
```

Comment:
Desperate attempt to improve the caches everywhere:
```
return edge().source().flatMap(edge().tx().factory()::buildConcept).orElseThrow(() ->
    GraknTxOperationException.missingOwner(getId())
);
```

(feel free to extract more variables to make it clearer or w/e)